horovod/mxnet/mpi_ops.py

```diff
@@ -158,6 +161,9 @@ def allgather(tensor, name=None, priority=0):
     else:
         check_call(MPI_MXNET_LIB_CTYPES.horovod_mxnet_allgather_async(
             c_in, c_out, name, ctypes.c_int(priority)))
+
+    # Need to block here so changes to output tensor are visible
+    output.wait_to_read()
```

Comment:
Is this necessary? WaitToRead was already called in line 219 in `mpi_ops.h`

Reply:
It is necessary for the regular `DoHorovodOperation` path, which is used for CPU and also for GPU if you install with `HOROVOD_GPU_ALLGATHER=NCCL`. In that path, there is no wait to read and as a result, you get the wrong output. For example, consider this script running with a build using `HOROVOD_GPU_ALLGATHER=NCCL`:
```
import mxnet as mx
import horovod.mxnet as hvd

hvd.init()

ctx = mx.gpu(hvd.local_rank())
tensor = mx.ndarray.ones(shape=[2], ctx=ctx) * hvd.rank()

gathered = hvd.allgather(tensor)

print(gathered)
``` 
The current PR produces the correct result:
```
$ mpirun -np 2 python test_mxnet_allgather.py

[0. 0. 1. 1.]
<NDArray 4 @gpu(0)>

[0. 0. 1. 1.]
<NDArray 4 @gpu(1)>
```
If I remove the `output.wait_to_read()`, it returns the incorrect result:
```
$ mpirun -np 2 --allow-run-as-root python test_mxnet_allgather.py
[5.002331e-36]
<NDArray 1 @gpu(0)>
[1.3273958e-34]
<NDArray 1 @gpu(1)>
```
Without the `wait_to_read`, it is just immediately returning the empty output, created here:
https://github.com/horovod/horovod/blob/032688871d04181d66d80cd2cfd50f3f4662073d/horovod/mxnet/mpi_ops.py#L154-L155

Trying to run the script using CPU input tensors without the `wait_to_read` results in an exception triggered by the final call to `print(gathered):
```
root@57a057477e3f:/opt/home/dl_multinode/horovod_fork/horovod_romerojosh/test/allgather# mpirun -np 2 --allow-run-as-root python test_mxnet_allgather.py
Traceback (most recent call last):
  File "test_mxnet_allgather.py", line 12, in <module>
    print(gathered)
  File "/opt/mxnet/python/mxnet/ndarray/ndarray.py", line 246, in __repr__
    return '\n%s\n<%s %s @%s>' % (str(self.asnumpy()),
  File "/opt/mxnet/python/mxnet/ndarray/ndarray.py", line 2535, in asnumpy
    ctypes.c_size_t(data.size)))
  File "/opt/mxnet/python/mxnet/base.py", line 255, in check_call
    raise MXNetError(py_str(_LIB.MXGetLastError()))
mxnet.base.MXNetError: [17:03:33] src/ndarray/ndarray_function.cc:56: Check failed: size == to->Size() (4 vs. 1) : copying size mismatch, from: 16 bytes, to:
4 bytes.
```
This issue is also fixed by having the `wait_to_read`.

Reply:
Thanks for the detailed explanation!